### PR TITLE
Provides the public part of metrics data collection and export

### PR DESCRIPTION
**Pull request by wxbty**
Since different developers implement the collection and processing of indicators in different scenarios, I think it is necessary to provide a unified API to handle the same parts, such as buried points, MerticsKey and other scenarios. This can also simplify the code review of the specific implementation part.

1. The initMetricsReporter of DefaultApplicationDeployerd obtains the collector through spi, and then registers with BeanFactory, and dynamically obtains the collector when exporting data. In this way, the default package does not need to rely on third-party collection packages, and users can customize collectors.
2. Add the MetricsCollector at the Application level to distinguish the method level.
3. Buried in the form of publishing events, adding a life cycle method for events, and describing the 3 fixed steps of the start, end, and exception of the buried stage.
4. TimeCounter integrates event and provides reusable event rt processing.
5. MetricsKeyWrapper can dynamically describe MetricsKey and reduce repetitive code of MetricsKey.
6. The data aggregation objects of each scene need to export the GaugeMetricSample format data in the end, and define the MetricsExport interface to describe the behavior of the same object.

中文：
由于不同的开发人员实现不同场景的指标收集处理，我认为有必要提供统一的api处理相同部分，比如埋点、MerticsKey等场景。这样可以有相似的代码风格，同时简化具体实现部分的代码review。
1、DefaultApplicationDeployerd的initMetricsReporter通过spi获取collector，再注册到BeanFactory，导出数据时动态获取collector。这样default包不需要依赖三方收集包，同时用户可以自定义collector。
2、增加Application级别的MetricsCollector，以区分method级别。
3、发布事件的形式进行埋点，为事件增加生命周期方法，描述埋点阶段开始、结束及异常3个固定步骤。
4、TimeCounter 集成event，提供可复用的事件rt处理。
5、MetricsKeyWrapper可以动态的描述MetricsKey，减少MetricsKey重复代码。
6、各场景的数据聚合对象，最终都需要导出GaugeMetricSample格式数据，定义MetricsExport接口描述同一对象的行为。

### CODE REVIEW

**Review by songxiaosheng — APPROVED** [review]
LGTM

**Review by songxiaosheng — APPROVED** [review]
LGTM

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java [hunk 1]
```diff
@@ -772,4 +774,26 @@ public Invoker<?> getInvoker() {
...
+    @Override
+    @Parameter(excluded = true, attribute = false)
+    public int getValidInvokerNum() {
+        Invoker<?> invoker = getInvoker();
+        if (invoker instanceof MigrationInvoker) {
+            AbstractDirectory<?> directory = (AbstractDirectory<?>) ((MigrationInvoker<?>) invoker).getDirectory();
+            return directory.getValidInvokers().size();
+        }
+        return super.getValidInvokerNum();
+    }
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/ReferenceConfig.java:798]: https://github.com/apache/dubbo/pull/11493#pullrequestreview-1296730732

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java [hunk 2]
```diff
@@ -357,13 +363,27 @@ private void initMetricsService() {
...
+            List<MetricsListener> metricsListeners = applicationModel.getExtensionLoader(MetricsListener.class)
+                .getActivateExtensions();
+            metricsListeners.forEach(this.eventMulticaster::addListener);
+
+            List<MetricsCollector> customizeCollectors = applicationModel.getExtensionLoader(MetricsCollector.class)
+                .getActivateExtensions();
+            for (MetricsCollector customizeCollector : customizeCollectors) {
+                beanFactory.registerBean(customizeCollector);
+            }
+
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java:386]: https://github.com/apache/dubbo/pull/11493#pullrequestreview-1296730732

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

**Comment by wxbty** [comment]
@AlbumenJ  PTAL  again

File: dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java [hunk 3]
```diff
@@ -358,9 +363,11 @@ private void initMetricsService() {
     }
 
     private void initMetricsReporter() {
-        DefaultMetricsCollector collector =
-            applicationModel.getFrameworkModel().getBeanFactory().getOrRegisterBean(DefaultMetricsCollector.class);
+        ScopeBeanFactory beanFactory = applicationModel.getFrameworkModel().getBeanFactory();
+        DefaultMetricsCollector collector = beanFactory.getOrRegisterBean(DefaultMetricsCollector.class);
         MetricsConfig metricsConfig = configManager.getMetrics().orElse(null);
+        beanFactory.registerBean(GlobalMetricsEventMulticaster.class);
+
```
Comment (AlbumenJ) [inline dubbo-config/dubbo-config-api/src/main/java/org/apache/dubbo/config/deploy/DefaultApplicationDeployer.java:370]: Move to `org.apache.dubbo.rpc.model.ScopeModelInitializer` like `org.apache.dubbo.config.ConfigScopeModelInitializer`, `org.apache.dubbo.registry.RegistryScopeModelInitializer` and `org.apache.dubbo.common.CommonScopeModelInitializer`.

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/event/GlobalMetricsEventMulticaster.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+public class GlobalMetricsEventMulticaster extends SimpleMetricsEventMulticaster {
+
+    public GlobalMetricsEventMulticaster(FrameworkModel frameworkModel) {
+        ScopeBeanFactory beanFactory = frameworkModel.getBeanFactory();
+        ExtensionLoader<MetricsCollector> extensionLoader = frameworkModel.getExtensionLoader(MetricsCollector.class);
+        if (extensionLoader != null) {
+            List<MetricsCollector> customizeCollectors = extensionLoader
+                .getActivateExtensions();
+            for (MetricsCollector customizeCollector : customizeCollectors) {
+                beanFactory.registerBean(customizeCollector);
```
Comment (AlbumenJ) [inline dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/event/GlobalMetricsEventMulticaster.java:36]: Why register to beanFactory?

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/event/GlobalMetricsEventMulticaster.java (hunk 4 above)
Comment (wxbty) [reply dubbo-metrics/dubbo-metrics-api/src/main/java/org/apache/dubbo/metrics/event/GlobalMetricsEventMulticaster.java:36]: Here (AbstractMetricsReporter.java) can be obtained dynamically,
was hardcoded before

**Review by AlbumenJ — APPROVED** [review]

**Pull request merged**